### ISSUE TITLE
switch to esm?

### ISSUE THREAD

**Issue by jimmywarting** [mapbox/node-pre-gyp#619]
Given that one of the dependency (node-fetch@3) is now esm.
How do you feel about updating the dependency and also switching to esm?

**Comment by springmeyer** [comment]
@jimmywarting - thanks for asking. I'm pretty unfamiliar with esm and the implications of a change like this. I presume it would mean we'd no longer be able to support older node versions, correct? And what would be the advantage of upgrading to node-fetch@3?

**Comment by jimmywarting** [comment]
> I presume it would mean we'd no longer be able to support older node versions, correct?

Correct, the least minium version node-fetch@3 support is node v12.20. but it can't be require'd like normal, you need to use ether `import fetch from 'node-fetch'` or `import('node-fetch').then(module => {})`, the later one works from cjs too

there is a solution that don't involve converting the entire project into ESM and making a drastic major breaking change to everyone, but that involves creating a small wrapper fn, something like: 
```js
const fetch = (...args) => import('node-fetch').then(({default: fetch}) => fetch(...args));
```

> And what would be the advantage of upgrading to node-fetch@3?

node-fetch@3 will get newer features and going to align more with the spec more and more
it will have a proper FormData, Blob & File class with whatwg readable streams, AbortController / AbortSignal all following the standard spec more correctly.
v4 will eventually change the response.body to a proper whatwg readable stream in some feature.

----

If you wish i can send you a PR that converts the entire project into ESM.

**Comment by springmeyer** [comment]
> If you wish i can send you a PR that converts the entire project into ESM.

👍🏼  thanks.

But I won't plan to merge it for some time because I don't currently have the bandwidth to maintain multiple branches/releases series. And I'd like to keep node v10 support around for another year (node-pre-gyp is depended upon by a lot of modules so it should not be the one that breaks first with older node versions)

**Comment by jimmywarting** [comment]
Hello again, would you mind if i moved some exports around?
mainly grouping them and putting them last (at the bottom of the file)

### reason for grouping them and having a single export

- An export declaration or `module.exports` assignment can appear anywhere in the code. By requiring a single export declaration all your exports will remain at one place, making it easier to see what exports a module provides.
- in the event that you would like to make the switch at some point later when switching to ESM then it would be just as easy with a single line replacement in each file that exports stuff. And it would be easier to review a such PR.
```diff
- module.exports = {
+ export {
    xyz
  }
```
The idea comes from: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/group-exports.md

### Reason for putting the export last at the bottom of the file.
- It dose not depend on things to be hoisted to the top of the file. Relying on `var` and `functions` to be hoisted is abnormal and not recommended in the event that someone wish to refactor a prototype based `function` into a class or a `const` or `let` which dose not automatically hoist things to the top. So in other words: refactoring to more modern code becomes easier to by having it at the bottom.
- It's unnecessary job for a engine / machine having to hoist things to the top. refactoring gets harder. the flow goes more like 
  1. "here is what i need to import"
  2. "here is my code"
  3. "This is what i need to export"
- Having the export last in all files is a common place to go to and look for to see what is being exported.
- just like grouping. it becomes easier to see what kind of exports a module provides. rather than having it scatter around all over the place 

This idea come from https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/exports-last.md

All this can be accomplish without changing any functionality of the code or breaking anything
